packages/core/src/shared/downloadLSP.ts

```diff
@@ -0,0 +1,214 @@
+            res.body.on('error', (err) => {
+                reject(err)
+            })
+            file.on('finish', () => {
+                file.close(resolve)
+            })
+        })
+    }
+
+    async getFileSha384(filePath: string): Promise<string> {
+        const fileBuffer = await fs.readFileBytes(filePath)
+        const hash = crypto.createHash('sha384')
+        hash.update(fileBuffer)
+        return hash.digest('hex')
+    }
+
+    private async hashMatch(filePath: string, content: Content) {
+        const sha384 = await this.getFileSha384(filePath)
+        if ('sha384:' + sha384 !== content.hashes[0]) {
+            getLogger('lsp').error(`Downloaded file sha ${sha384} does not match manifest ${content.hashes[0]}.`)
+            await fs.delete(filePath)
+            return false
+        }
+        return true
+    }
+
+    async downloadAndCheckHash(filePath: string, content: Content) {
+        await this._download(filePath, content.url)
+        const match = await this.hashMatch(filePath, content)
+        if (!match) {
+            return false
+        }
+        return true
+    }
+
+    getDependency(manifest: Manifest, name: string): Content | undefined {
+        if (manifest.isManifestDeprecated) {
+            return undefined
+        }
+        for (const version of manifest.versions) {
+            if (version.isDelisted) {
+                continue
+            }
+            if (this.supportedLspServerVersions && !this.supportedLspServerVersions.includes(version.serverVersion)) {
+                continue
+            }
+            for (const t of version.targets) {
+                if (
+                    (t.platform === process.platform || (t.platform === 'windows' && process.platform === 'win32')) &&
+                    t.arch === process.arch
+                ) {
+                    for (const content of t.contents) {
+                        if (content.filename.startsWith(name) && content.hashes.length > 0) {
+                            content.serverVersion = version.serverVersion
+                            return content
+                        }
+                    }
+                }
+            }
+        }
+        return undefined
+    }
+
+    async downloadAndExtractServer({
+        content,
+        installLocation,
+        name,
+        tempFolder,
+        extractToTempFolder = false,
+    }: {
+        content: Content
+        installLocation: string
+        name: string
+        tempFolder: string
+        extractToTempFolder?: boolean
+    }) {
+        const serverZipTempPath = path.join(tempFolder, `${name}.zip`)
+        const downloadOk = await this.downloadAndCheckHash(serverZipTempPath, content)
+        if (!downloadOk) {
+            return false
+        }
+
+        // load the zip contents
+        const extractPath = extractToTempFolder ? tempFolder : path.join(tempFolder, name)
+        new AdmZip(serverZipTempPath).extractAllTo(extractPath)
```

Comment:
While we're here, we should try to migrate to the new `shared/utilities/zipStream.ts` if possible. AdmZip will be removed (not web-compatible; though may be irrelevant for this module(?) but we want to eliminate AdmZip totally to avoid confusion about when to use which)

Reply:
:+1: i'll make a followup PR to change that